zookeeper-server/src/main/java/org/apache/zookeeper/ClientCnxn.java

```diff
@@ -1121,6 +1125,9 @@ private void startConnect(InetSocketAddress addr) throws IOException {
                     LOG.warn("Unexpected exception", e);
                 }
             }
+            if (!state.isAlive()) {
```

Comment:
I suspect this will not completely fix the race condition between setting the state to closed and connecting. Another thread could set the state to close immediately after this check but before the subsequent `state = States.CONNECTING;` statement. We could make sure that this "check and assign" operation is atomic to completely address the race condition.

Reply:
Thanks for your comments. I agree with your concern. I would add synchronized(state) to ensure the atomic operation. I also add synchronized(state) guard at SendThread.close() to set state as States.CLOSED. I think these guards can resolve the problem.

Reply:
synchronized(state) is still not safe. Maybe SendThread.close() and SendThread.startConnect() synchronize on different objects because state is mutable and can be changed between SendThread.close() and SendThread.startConnect().
I prefer synchronized(this) guard.

Reply:
I think this will work. 

I would recommend instead of doing the check here, we add a set of new `synchronized` methods to `SenderThread` where we get and set the states of the `SenderThread.state`. Then, we can move the checks in the `setState` method, where we can put constraints on how the connection state may transit. For example, we can enforce in that method that a `CLOSED` state can't be transit to a different state, which will prevent the problem we are seeing in this issue (that setting a state to `CONNECTING` after being in `CLOSED` state). This has additional benefits as currently, there are multiple write paths to the `SenderThread.state`, and by consolidate read and write access of the state in these synchronized methods, we can be sure there will not be future race conditions w.r.t accessing this state variable.

I also thought about an alternative - instead of synchronizing on `SenderThread`, we could wrap the state variable in an `AtomicReference` and then provide get/set/compareAndSet methods on top of the atomic reference. This however is less flexible to allow us to control how the state could transit, so it still looks to me that  synchronizing on `SenderThread` is better.

Reply:
Accept it. Now I make a new method SendThread.setZkState() with checked IOException. Please check the latest patch.